### DISCUSSION TITLE
Netbox LDAP authentication class error - "LDAPbackend" attribute/class

### DISCUSSION THREAD

**Discussion by berkeleycowboy** [netbox-community/netbox#14067] [Q&A]
Receiving the following error accessing the Netbox server after configuring configuration.py and ldap_config.py

     <class 'ImportError'>

     Module "netbox.authentication" does not define a "LDAPbackend" attribute/class

configuration.py:

      REMOTE_AUTH_ENABLED = True
      REMOTE_AUTH_BACKEND = 'netbox.authentication.LDAPbackend'

Installation of the django-auth-ldap was successful.  Is there a particular folder the installation should be performed in ?  

[root@scdl-ifipam-001 netbox]# pwd
/opt/netbox/netbox
[root@scdl-ifipam-001 netbox]# source /opt/netbox/venv/bin/activate
(venv) [root@scdl-ifipam-001 netbox]# pip3 install django-auth-ldap
Requirement already satisfied: django-auth-ldap in /opt/netbox-3.5.8/venv/lib/python3.9/site-packages (4.6.0)
Requirement already satisfied: Django>=3.2 in /opt/netbox-3.5.8/venv/lib/python3.9/site-packages (from django-auth-ldap) (4.1.10)
Requirement already satisfied: python-ldap>=3.1 in /opt/netbox-3.5.8/venv/lib/python3.9/site-packages (from django-auth-ldap) (3.4.3)
Requirement already satisfied: asgiref<4,>=3.5.2 in /opt/netbox-3.5.8/venv/lib/python3.9/site-packages (from Django>=3.2->django-auth-ldap) (3.7.2)
Requirement already satisfied: sqlparse>=0.2.2 in /opt/netbox-3.5.8/venv/lib/python3.9/site-packages (from Django>=3.2->django-auth-ldap) (0.4.4)
Requirement already satisfied: pyasn1>=0.3.7 in /opt/netbox-3.5.8/venv/lib/python3.9/site-packages (from python-ldap>=3.1->django-auth-ldap) (0.5.0)
Requirement already satisfied: pyasn1-modules>=0.1.5 in /opt/netbox-3.5.8/venv/lib/python3.9/site-packages (from python-ldap>=3.1->django-auth-ldap) (0.3.0)
Requirement already satisfied: typing-extensions>=4 in /opt/netbox-3.5.8/venv/lib/python3.9/site-packages (from asgiref<4,>=3.5.2->Django>=3.2->django-auth-ldap) (4.7.1)

[notice] A new release of pip is available: 23.2.1 -> 23.3
[notice] To update, run: pip install --upgrade pip

**Comment by candlerb** [comment]
`netbox.authentication.LDAPbackend` is not the same as `netbox.authentication.LDAPBackend` (capital 'B' required)

**Comment by berkeleycowboy** [comment]
Thank you.  Forgot about Linux capitalization.   

Still no joy.  Now complaining about "No module named 'django_auth_ldap.backend'"

Netbox Web interface output with debugging enabled:

[ModuleNotFoundError at _.pdf](https://github.com/netbox-community/netbox/files/13062241/ModuleNotFoundError.at._.pdf)

Do the OpenLDAP libraries need to be installed ?    This document indicates they are required.  

https://django-auth-ldap.readthedocs.io/en/latest/install.html

This site provides a method for Installing python-ldap.  Does anybody have any experience installing this in the virtual environment.  

https://www.python-ldap.org/en/python-ldap-3.3.0/installing.html

**Comment by berkeleycowboy** [comment]
The Exception Location references  /opt/netbox/netbox/netbox/authentication.py, line 319, in __new__
Code Fragment:
315  class LDAPBackend:
   316
   317      def __new__(cls, *args, **kwargs):
   318          try:
   319              from django_auth_ldap.backend import LDAPSettings
   320              import ldap
   321          except ModuleNotFoundError as e:
   322              if getattr(e, 'name') == 'django_auth_ldap':
   323                  raise ImproperlyConfigured(
   324                      "LDAP authentication has been configured, but django-auth-ldap is not installed."
   325                  )

The following command returns the following:
(venv) [root@scdl-ifipam-001 django_auth_ldap]# egrep "^class" -n /opt/netbox/netbox/netbox/authentication.py /opt/netbox-3.5.8/venv/lib/python3.9/site-packages/django_auth_ldap/backend.py
/opt/netbox/netbox/netbox/authentication.py:63:class ObjectPermissionMixin:
/opt/netbox/netbox/netbox/authentication.py:138:class ObjectPermissionBackend(ObjectPermissionMixin, ModelBackend):
/opt/netbox/netbox/netbox/authentication.py:142:class RemoteUserBackend(_RemoteUserBackend):
/opt/netbox/netbox/netbox/authentication.py:315:class LDAPBackend:
/opt/netbox-3.5.8/venv/lib/python3.9/site-packages/django_auth_ldap/backend.py:81:class LDAPBackend:
/opt/netbox-3.5.8/venv/lib/python3.9/site-packages/django_auth_ldap/backend.py:247:class _LDAPUser:
/opt/netbox-3.5.8/venv/lib/python3.9/site-packages/django_auth_ldap/backend.py:866:class _LDAPUserGroups:

Why is LDAPBackend defined in both files ?   
     /opt/netbox/netbox/netbox/authentication.py
     /opt/netbox-3.5.8/venv/lib/python3.9/site-packages/django_auth_ldap/backend.py

Code Fragment of /opt/netbox/netbox/netbox/authentication.py
    315 class LDAPBackend:
    316 
    317     def __new__(cls, *args, **kwargs):
    318         try:
    319             from django_auth_ldap.backend import LDAPSettings
    320             import ldap
    321         except ModuleNotFoundError as e:
    322             if getattr(e, 'name') == 'django_auth_ldap':
    323                 raise ImproperlyConfigured(
    324                     "LDAP authentication has been configured, but django-auth-ldap is not installed."
    325                 )
    326             raise e
    327 

Code Fragment of /opt/netbox-3.5.8/venv/lib/python3.9/site-packages/django_auth_ldap/backend.py
     81 class LDAPBackend:
     82     """
     83     The main backend class. This implements the auth backend API, although it
     84     actually delegates most of its work to _LDAPUser, which is defined next.
     85     """
     86 
     87     supports_anonymous_user = False
     88     supports_object_permissions = True
     89     supports_inactive_user = False
     90 
     91     _settings = None
     92     _ldap = None  # The cached ldap module (or mock object)
     93

**Reply by abhi1693** [reply]
You need to install django-auth-ldap

**Comment by berkeleycowboy** [comment]
Thanks for the reply.  

Yes,  django-auth-ldap has successfully been installed in the Python virtual environment.

(venv) [root@scdl-ifipam-001 site-packages]# pip3 install django-auth-ldap
Requirement already satisfied: django-auth-ldap in /opt/netbox-3.5.8/venv/lib/python3.9/site-packages (4.6.0)
Requirement already satisfied: Django>=3.2 in /opt/netbox-3.5.8/venv/lib/python3.9/site-packages (from django-auth-ldap) (4.1.10)
Requirement already satisfied: python-ldap>=3.1 in /opt/netbox-3.5.8/venv/lib/python3.9/site-packages (from django-auth-ldap) (3.4.3)
Requirement already satisfied: asgiref<4,>=3.5.2 in /opt/netbox-3.5.8/venv/lib/python3.9/site-packages (from Django>=3.2->django-auth-ldap) (3.7.2)
Requirement already satisfied: sqlparse>=0.2.2 in /opt/netbox-3.5.8/venv/lib/python3.9/site-packages (from Django>=3.2->django-auth-ldap) (0.4.4)
Requirement already satisfied: pyasn1>=0.3.7 in /opt/netbox-3.5.8/venv/lib/python3.9/site-packages (from python-ldap>=3.1->django-auth-ldap) (0.5.0)
Requirement already satisfied: pyasn1-modules>=0.1.5 in /opt/netbox-3.5.8/venv/lib/python3.9/site-packages (from python-ldap>=3.1->django-auth-ldap) (0.3.0)
Requirement already satisfied: typing-extensions>=4 in /opt/netbox-3.5.8/venv/lib/python3.9/site-packages (from asgiref<4,>=3.5.2->Django>=3.2->django-auth-ldap) (4.7.1)

Defined Classes in authentication.py and backend.py :

(venv) [root@scdl-ifipam-001 site-packages]# egrep "^class" -n /opt/netbox/netbox/netbox/authentication.py 
63:class ObjectPermissionMixin:
138:class ObjectPermissionBackend(ObjectPermissionMixin, ModelBackend):
142:class RemoteUserBackend(_RemoteUserBackend):
315:class LDAPBackend:

(venv) [root@scdl-ifipam-001 site-packages]# egrep "^class" -n  /opt/netbox-3.5.8/venv/lib/python3.9/site-packages/django_auth_ldap/backend.py
81:class LDAPBackend:
247:class _LDAPUser:
866:class _LDAPUserGroups:

Other steps that have been confirmed:
1.  https://django-auth-ldap.readthedocs.io/en/latest/
2.  https://www.openldap.org/ => confirmed openldap-2.4 has been installed

This is the offending command that's reported when debugging is enabled:

319             from django_auth_ldap.backend import LDAPSettings

django_auth_ldap.backend appears to be defined but where is LDAPSettings defined ?  

debugging information:

[ModuleNotFoundError.at._.pdf](https://github.com/netbox-community/netbox/files/13329306/ModuleNotFoundError.at._.pdf)

Server Components:
REHL 8.8
Netbox Version 3.5.8
Python Version:  3.9
PostgresSQL version:  15.3
Redis:  5.0.3
django-auth-ldap==4.6.0  ; from server PIP freeze in Pyton virtual Environment.  

Have compared the Pip freeze versions on the Netbox server to what is [published](https://github.com/markkuleinio/netbox-freezes/blob/main/netbox-3.5.8.txt) 
[pip freeze compare.pdf](https://github.com/netbox-community/netbox/files/13329933/pip.freeze.compare.pdf)

Thanks in advance for any troubleshooting or guidance.

**Reply by t0ta11ed** [reply]
I'm seeing this missing module error as well on a fresh Netbox install. The module was installed in the virtual environment but returns the error in the browser

No module named django_auth_ldap.backend

Python 3.10 and NetBox v 3.6.6

**Reply by berkeleycowboy** [reply]
Thank you t0ta11ed for your reply.   

Validates that I'm followed the recipe correctly.

**Comment by berkeleycowboy** [comment]
I've been trying to get Netbox running Redhat for approximately a year.   Have not been able to get Secure LDAP working.  

I have heard that Netbox Binary images is a posibility.  

Where can I find this a binary netbox image ?

**Reply by DanSheps** [reply]
There are no binary images. NetBox is 100% a python application, and hence no binary (other then Python).

I am not sure what problems you are having with secure LDAP, but you could always post your config (sanitized of course) and people could try and help you.

Other options include:

* NetBox Cloud (Cloud offering by NetBox Labs) w/ SSO (Azure AD, Okta, etc)
* SSO configuration instead of LDAP (I personally have moved to an "OpenID Connect" (Duo actually) based SSO)
* You could also hire someone (there are likely a few people who would be able to assist you both here and in slack)

**Reply by candlerb** [reply]
There is also the [docker image](https://github.com/netbox-community/netbox-docker). However, you still need to tailor it for [LDAP](https://github.com/netbox-community/netbox-docker/wiki/LDAP) or other [special requirements](https://github.com/netbox-community/netbox-docker/wiki/Configuration#custom-configuration-files).

**Reply by DanSheps** [reply]
Yes, sorry, absolutely correct, there definitely is a docker image

**Reply by jcollie** [reply]
Other than configuration, the Docker image doesn't need anything special for LDAP. Installing Python apps "from scratch" can be very tricky to get all of the dependencies right. It's definitely the way to go to eliminate installation problems related to Python and related libraries.
